json-to-messages: catch JSON parsing errors

<!-- NAMING YOUR ISSUE: Please prefix your issue with the name of the sub-project -->
<!-- e.g. `messages: Define new StepDef message` -->
<!-- This makes it easier to get some context when reading the names of issues -->

<!-- These sections are meant as guidance for you, to help you give the kind of information we'll need to help with your issue. If a section  doesn't seem to fit, just skip it.

In general: Please provide as much information as you can to help us solving your problem -->

## Summary

I'm trying to use `json-to-messages` as a library but I'm unable to trap JSON parsing errors, e.g.

```js
const readme = fs.createReadStream('README.md')
jsonToMessages(readme, process.stdout).
  catch(err => { console.error(`Trapped error: ${err}`) })
```

## Expected Behavior

<!--- If you're describing a bug, tell us what should happen -->
<!--- If you're suggesting a change/improvement, tell us how it should work -->
<!--- Feel free to use Given / Then / Then if that helps, but please add some plain-language context too -->

I would expect the promise returned from `jsonToMessages` to be rejected with the error.

## Current Behavior

<!--- If describing a bug, tell us what happens instead of the expected behavior -->
<!--- If suggesting a change/improvement, explain the difference from current behavior -->

<!--- If you have got some output place it in the code block below. Otherwise remove it. -->
~~~
$ npm start

> node server.js

undefined:1
# Setting up your development environment
^

SyntaxError: Unexpected token # in JSON at position 0
    at JSON.parse (<anonymous>)
    at JSONTransformStream._flush (/Users/tooky/Code/json-to-messages-service/node_modules/@cucumber/json-to-messages/dist/src/stream/JSONTransformStream.js:30:24)
    at JSONTransformStream.prefinish (_stream_transform.js:140:10)
    at JSONTransformStream.emit (events.js:210:5)
    at prefinish (_stream_writable.js:645:14)
    at finishMaybe (_stream_writable.js:653:5)
    at endWritable (_stream_writable.js:673:3)
    at JSONTransformStream.Writable.end (_stream_writable.js:604:5)
    at IncomingMessage.onend (_stream_readable.js:691:10)
    at Object.onceWrapper (events.js:299:28)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! json-to-messages-service@0.0.1 start: `node server.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the json-to-messages-service@0.0.1 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/tooky/.npm/_logs/2020-05-07T08_23_54_390Z-debug.log
 $ echo $?
1
~~~

An error is thrown internally but it exits the whole node process with error status 1.

--- a/json-to-messages/CHANGELOG.md
+++ b/json-to-messages/CHANGELOG.md
@@ -17,6 +17,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 
 ### Fixed
 
+* Catch JSON parsing errors [#992](https://github.com/cucumber/cucumber/issues/992) (@tooky)
+
 ## [1.0.1] - 2020-04-22
 
 ### Added

--- a/json-to-messages/javascript/src/stream/JSONTransformStream.ts
+++ b/json-to-messages/javascript/src/stream/JSONTransformStream.ts
@@ -18,7 +18,11 @@ export default class JSONTransformStream extends Transform {
   }
 
   _flush(callback: TransformCallback): void {
-    this.push(JSON.parse(this.stringChunks.join('')))
-    callback()
+    try {
+      this.push(JSON.parse(this.stringChunks.join('')))
+      callback()
+    } catch (err) {
+      callback(err)
+    }
   }
 }

--- a/json-to-messages/javascript/test/stream/JSONTransformStreamTest.ts
+++ b/json-to-messages/javascript/test/stream/JSONTransformStreamTest.ts
@@ -30,7 +30,7 @@ describe('JSONReportStream', () => {
       },
     })
     const sink = new SingleObjectWritableStream<Array<any>>()
-    let f = () => {}
+    let f = (): null => null
     try {
       await asyncPipeline(readableStream, new JSONTransformStream(), sink)
     } catch (e) {

--- a/json-to-messages/javascript/test/stream/JSONTransformStreamTest.ts
+++ b/json-to-messages/javascript/test/stream/JSONTransformStreamTest.ts
@@ -19,4 +19,26 @@ describe('JSONReportStream', () => {
     await asyncPipeline(readableStream, new JSONTransformStream(), sink)
     assert.deepStrictEqual(sink.object, data)
   })
+
+  it('handles non-json data gracefully', async () => {
+    const data = '# this is not json'
+
+    const readableStream = new Stream.Readable({
+      read() {
+        this.push(data)
+        this.push(null)
+      },
+    })
+    const sink = new SingleObjectWritableStream<Array<any>>()
+    let f = () => {}
+    try {
+      await asyncPipeline(readableStream, new JSONTransformStream(), sink)
+    } catch (e) {
+      f = () => {
+        throw e
+      }
+    } finally {
+      assert.throws(f, /SyntaxError.+JSON/)
+    }
+  })
 })

